cmd/kueueviz/backend/main.go

```diff
@@ -20,36 +20,120 @@ import (
 	"fmt"
 	"log"
 	"net/http"
+	"net/url"
+	"strings"
+	"time"
 
+	"github.com/gin-contrib/cors"
 	"github.com/gin-gonic/gin"
 	"github.com/spf13/viper"
 	"kueueviz/handlers"
 
 	_ "net/http/pprof"
 )
 
+// validateOrigin validates and sanitizes a CORS origin URL
+func validateOrigin(origin string) (string, bool) {
+	u, err := url.Parse(origin)
+	if err != nil {
+		return "", false
+	}
+
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return "", false
+	}
+
+	if u.Hostname() == "" {
+		return "", false
+	}
+
+	cleanOrigin := fmt.Sprintf("%s://%s", u.Scheme, u.Host)
+	return cleanOrigin, true
+}
+
+// configureCORS sets up CORS configuration based on environment
+func configureCORS() cors.Config {
+	config := cors.DefaultConfig()
+
+	// Get allowed origins from environment variable
+	allowedOriginsEnv := viper.GetString("KUEUEVIZ_ALLOWED_ORIGINS")
+	var allowedOrigins []string
+
+	if allowedOriginsEnv != "" {
+		// Split comma-separated origins and validate each
+		rawOrigins := strings.Split(allowedOriginsEnv, ",")
+		for _, origin := range rawOrigins {
+			origin = strings.TrimSpace(origin)
+			if cleanOrigin, valid := validateOrigin(origin); valid {
+				allowedOrigins = append(allowedOrigins, cleanOrigin)
+			} else {
+				log.Printf("Warning: Invalid origin '%s' rejected", origin)
+			}
+		}
+	} else {
+		// Default development origins (only in development mode)
+		if gin.Mode() != gin.ReleaseMode {
+			allowedOrigins = []string{
+				"http://localhost:3000",
+				"http://127.0.0.1:3000",
+			}
+			log.Println("KUEUEVIZ_ALLOWED_ORIGINS not set, using default development origins")
+		} else {
+			log.Println("Production mode: KUEUEVIZ_ALLOWED_ORIGINS must be explicitly set")
+		}
+	}
+
+	config.AllowOrigins = allowedOrigins
+	config.AllowMethods = []string{"GET", "HEAD", "OPTIONS"}
+	config.AllowHeaders = []string{"Origin", "Content-Length", "Content-Type", "Authorization"}
+
+	if gin.Mode() == gin.ReleaseMode {
+		log.Println("Running in production mode, applying security settings")
+		config.AllowCredentials = false
+		config.MaxAge = 12 * time.Hour
+
+		log.Printf("CORS allowed origins: %v", config.AllowOrigins)
+
+		if len(allowedOrigins) == 0 {
+			log.Fatal("Production mode requires valid CORS origins to be configured")
+		}
+	} else {
+		log.Println("Running in development mode")
+		config.AllowCredentials = true
+		config.MaxAge = 5 * time.Minute
+	}
+	return config
+}
+
 func main() {
 	viper.AutomaticEnv()
-	// Start pprof server for profiling
-	go func() {
-		log.Println("Starting pprof server on :6060")
-		if err := http.ListenAndServe("localhost:6060", nil); err != nil {
-			log.Fatalf("Error starting pprof server: %v", err)
-		}
-	}()
+	if gin.Mode() != gin.ReleaseMode {
```

Comment:
I’m wondering if we need to add GIN_MODE variable to the manifests.

Reply:
👌

Reply:
When building the YAML manifests for the release, I think we need to add this variable – perhaps [here](https://github.com/kubernetes-sigs/kueue/blob/50e4ba6a497298b1a5523f62ebbb9d5da85422fb/config/kueueviz/kustomization.yaml). But this is not a related change, so we can move it to a follow-up PR.